torchdata/datapipes/iter/transform/callable.py

```diff
@@ -168,6 +171,140 @@ def __len__(self) -> int:
         raise TypeError(f"{type(self).__name__}'s length relies on the output of its function.")
 
 
+@functional_datapipe("shuffled_flatmap")
+class ShuffledFlatMapperIterDataPipe(IterDataPipe):
+    r"""
+    Applies a function over each item from the source DataPipe,
+    then collects the iterables returned in a buffer,
+    then, at every iteration, chooses at random one of the iterables in the buffer
+    and yields one item from this iterable (functional name: ``shuffled_flatmap``).
+
+    Note:
+        The output from ``fn`` must be an Iterable. Otherwise, an error will be raised.
+        If ``fn`` is ``None``, source DataPipe will be just flattened vertically, provided that items can be unpacked.
+
+    Args:
+        datapipe: Source IterDataPipe
+        fn: the function to be applied to each element in the DataPipe, the output must be a Sequence
+        input_col: Index or indices of data which ``fn`` is applied, such as:
+
+            - ``None`` as default to apply ``fn`` to the data directly.
+            - Integer(s) is/are used for list/tuple.
+            - Key(s) is/are used for dict.
+        buffer_size: The buffer size for sampling the iterables (default to ``10000``)
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> source_dp = IterableWrapper([[1, 2, 3, 4], 'abcd', 'ABCD'])
+        >>> shuffled_flatmapped_dp = source_dp.shuffled_flatmap(buffer_size=2)
+        >>> list(shuffled_flatmapped_dp)
+        ['a', 'b', 'c', 1, 'd', 'A', 'B', 'C', 2, 'D', 3, 4]
+    """
+    datapipe: IterDataPipe[T_co]
+    fn: Optional[Callable]
+    buffer_size: int
+    _enabled: bool
+    _seed: Optional[int]
+    _rng: random.Random
+
+    def __init__(
+        self, datapipe: IterDataPipe, fn: Optional[Callable] = None, input_col=None, buffer_size: int = 10000
```

Comment:
> This default `buffer_size` might too big to me. If the output of `flatmap` is like 1k, the total buffer size becomes 10,000,000. Can we add two sizes of buffer? One is the size of input data, one is the total size of buffer `sum(len(buffer) for buffer in buffers)`

Yes, `buffer_size=10000` is probably too big. what about 1000?

Reply:
> Can we remove `input_col` for this operation for now? It makes it too complicated. And, we can easily replace the operation with `input_col` to `dp.drop(columns).map(map_fn)`.

Agree entirely, I kept `input_col` because it was present in `FlatMapper`. Tbh, I would also get rid of the `fn` argument because the "map" part can be achieved with a `Mapper` datapipe before. However, the way I flatten the inputs is closer to the behaviour of `FlatMapper` than of `Flattener`.

By the way, the names `FlatMapper` and `Flattener` can be confusing because they don't have the same purpose...

Reply:
> > This default `buffer_size` might too big to me. If the output of `flatmap` is like 1k, the total buffer size becomes 10,000,000. Can we add two sizes of buffer? One is the size of input data, one is the total size of buffer `sum(len(buffer) for buffer in buffers)`
> 
> Yes, `buffer_size=10000` is probably too big. What about 1000?

Let's do 100 for now? There is one metric like the time for the first batch/element, which is also important. Making buffer size big causes longer time to get the first batch.

Yea, since you have already implemented, let's keep it as it is. I need to take a deeper look at if removing `input_col` from `flatmap` and this operation is doable by combining with other DataPipes like `drop`.